### ISSUE TITLE
why all processes get allocated on 1 gpu?

### ISSUE THREAD

**Issue by harryxu-yscz** [uber/horovod#240]
As title suggests. And here is what I observe from `nvidia-smi`. The command I used was `mpirun -np 2 python run.py <flags>`

```
|    0     26107      C   python                                      3587MiB |
|    0     26108      C   python                                      4425MiB |
|    1     26107      C   python                                       169MiB |
|    1     26108      C   python                                       169MiB |
```

**Comment by alsrgv** [comment]
@harryxu-yscz, I suspect that you missed setting https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py#L101 or you have some TensorFlow code that creates a session without this config (e.g. if you're enumerating all the available GPUs).

**Comment by harryxu-yscz** [comment]
Thanks for the reply!

In fact I did include that line. Another weird behavior I observed is that all my logging statements are print out twice..

```
2018-04-11 16:00:30,745 - brc - INFO - Dev set size: 5000 questions.
2018-04-11 16:00:30,745 - brc - INFO - Converting text into ids...
2018-04-11 16:00:30,789 - brc - INFO - Dev set size: 5000 questions.
2018-04-11 16:00:30,789 - brc - INFO - Converting text into ids...
2018-04-11 16:00:35,341 - brc - INFO - Initialize the model...
2018-04-11 16:00:35.346769: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-04-11 16:00:35.555385: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1344] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:02:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
2018-04-11 16:00:35,651 - brc - INFO - Initialize the model...
2018-04-11 16:00:35.662218: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-04-11 16:00:35.813071: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1344] Found device 1 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.7335
pciBusID: 0000:03:00.0
totalMemory: 7.92GiB freeMemory: 7.80GiB
```

**Comment by harryxu-yscz** [comment]
I solved the first issue: processes get placed on single gpu. Turned out I was not running the horovod part of the code..

Now my remaining question is the double printing. It looks like each gpu training process is replicating the training behavior: create loggings, generating data from its own generator, evaluating results. Any suggestion on this?

**Comment by alsrgv** [comment]
@harryxu-yscz, we did an internal comparison between data loader that does deterministic partitioning where there's no overlap between workers in a single epoch, and completely random data loader which reads random slices from the whole dataset with potential for some examples to be read multiple times in a single epoch.  We found no observable impact on the convergence.  So, the recommendation is to do whatever is easier / more beautiful.
